### PR TITLE
Enable doc size tracking

### PR DESCRIPTION
**Pull request by roonyh**

### CODE REVIEW

File: lib/docsize_cache.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  var key = this.getKey(coll, query, opts);
+  var item = this.items.get(key);
+
+  if (!item) {
+    item = new DocSzCacheItem(this.maxValues);
+    this.items.set(key, item);
+  }
+
+  if (this.needsUpdate(item)) {
+    var size = JSON.stringify(data[0]).length;
```
Comment (arunoda) [inline lib/docsize_cache.js]: Are we only getting the 0th element only?
Can we improve this for larger arrays. May be to take few items.

CC/ @mnmtanish

File: lib/docsize_cache.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    item.addData(size);
+  }
+
+  return item.getValue();
+};
+
+DocSzCache.prototype.getKey = function (coll, query, opts) {
+  var keyData = JSON.stringify([coll, query, opts]);
+  return crypto.createHash('md5').update(keyData).digest('hex');
```
Comment (arunoda) [inline lib/docsize_cache.js]: Do not use crypto. Just return the stringified value. Which is `keyData`

File: lib/docsize_cache.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  var currentTime = Date.now();
+  var timeSinceUpdate = currentTime - item.updated;
+  if (timeSinceUpdate > 1000*60) {
+    return true;
+  }
+
+  var cpuUsage = item.getPcpu();
+  if (cpuUsage > 80) {
```
Comment (arunoda) [inline lib/docsize_cache.js]: Even if it's greater then 80%, make sure to get some value.
But this times, it's much lower rate than earlier.

File: lib/docsize_cache.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    (100 - this.getPcpu()) / 100,
+  ].map(function (score) {
+    return score > 1 ? 1 : score;
+  }).reduce(function (total, score) {
+    return (total || 0) + score;
+  }) / 3;
+};
+
+DocSzCacheItem.prototype.getPcpu = function () {
+  var model = Kadira.models.system;
```
Comment (arunoda) [inline lib/docsize_cache.js]: Here we need to provide an API to update the CPU usage.

File: lib/hijack/wrap_observers.js [hunk 5]
```diff
@@ -1,13 +1,30 @@
 wrapOplogObserveDriver = function(proto) {
-  // Track the polled documents. This is reflect to the RAM size and 
+  // Track the polled documents. This is reflect to the RAM size and
   // for the CPU usage directly
   var originalPublishNewResults = proto._publishNewResults;
   proto._publishNewResults = function(newResults, newBuffer) {
+    var data = [];
+    newResults.forEach(function(item){
```
Comment (arunoda) [inline lib/hijack/wrap_observers.js]: Don't loop this. Find some way to send these two values to the `getSize()`  at once. Like:

~~~js
var docSize = Kadira.docSzCache.getSize(coll, query, opts, newResults, newBuffer);
~~~

And modify getSize to work with IDmaps too.

With that, we can avoid some loops.

File: lib/models/pubsub.js [hunk 6]
```diff
@@ -136,7 +136,9 @@ PubsubModel.prototype._getMetrics = function(timestamp, publication) {
       initiallyAddedDocuments: 0,
       liveAddedDocuments: 0,
       liveChangedDocuments: 0,
-      liveRemovedDocuments: 0
+      liveRemovedDocuments: 0,
+      InitiallyFetchedDocSize: 0,
+      LiveFetchedDocSize: 0
```
Comment (arunoda) [inline lib/models/pubsub.js]: We need another field for polledDocSize too.

File: lib/models/pubsub.js [hunk 7]
```diff
@@ -327,4 +329,19 @@ PubsubModel.prototype.trackLiveUpdates = function(info, type, count) {
...
+PubsubModel.prototype.trackDocSize = function(info, type, size) {
+  var timestamp = Ntp._now();
+  var publication = this._getMetrics(timestamp, info.name);
+
+  if(type === "polledFetches") {
+    publication.polledFetchedDocSize += size;
+  } else if(type === "liveFetches") {
+    publication.LiveFetchedDocSize += size;
+  } else if(type === "initialFetches") {
+    publication.initiallyAddedDocuments += size;
```
Comment (arunoda) [inline lib/models/pubsub.js]: This should be: `InitiallyFetchedDocSize`. Right?

File: lib/models/pubsub.js [hunk 8]
```diff
@@ -327,4 +329,19 @@ PubsubModel.prototype.trackLiveUpdates = function(info, type, count) {
...
     throw new Error("Kadira: Unknown live update type");
   }
+}
+
+PubsubModel.prototype.trackDocSize = function(info, type, size) {
+  var timestamp = Ntp._now();
+  var publication = this._getMetrics(timestamp, info.name);
+
+  if(type === "polledFetches") {
+    publication.polledFetchedDocSize += size;
```
Comment (arunoda) [inline lib/models/pubsub.js]: We need to add this to the line #141 right?

File: lib/hijack/db.js [hunk 9]
```diff
@@ -164,7 +164,14 @@ hijackDBOps = function hijackDBOps() {
...
         } else if(type == 'fetch' || type == 'map'){
           //for other cursor operation
+
           endData.docsFetched = ret.length;
+
+          var coll = cursorDescription.collectionName;
+          var query = cursorDescription.selector;
+          var opts = cursorDescription.options;
+          var docSize = Kadira.docSzCache.getSize(coll, query, opts, ret);
+          endData.docSize = docSize * ret.length;
```
Comment (arunoda) [inline lib/hijack/db.js]: Here we need to track these and cursorDocSize. Which should go to either the method or publication depending on the trace type.

File: lib/docsize_cache.js (hunk 1 above)
Comment (mnmtanish) [inline lib/docsize_cache.js]: Shall we use the mean value of all items? I think using the mean will be fair in this case and calculating the mean is also quite easy (stringify the whole array and divide by array length).

File: lib/docsize_cache.js (hunk 2 above)
Comment (mnmtanish) [inline lib/docsize_cache.js]: On it

Comment (mnmtanish) [inline lib/docsize_cache.js]: Done

File: lib/docsize_cache.js (hunk 3 above)
Comment (mnmtanish) [inline lib/docsize_cache.js]: Done

File: lib/docsize_cache.js (hunk 4 above)
Comment (mnmtanish) [inline lib/docsize_cache.js]: Doen

File: lib/docsize_cache.js (hunk 1 above)
Comment (arunoda) [inline lib/docsize_cache.js]: As we discussed on this matter, pick first few values based on the number of items in the array.

File: lib/models/methods.js [hunk 10]
```diff
@@ -30,6 +30,37 @@ MethodsModel = function (metricsThreshold) {
...
+  }
+
+  var methods = this.methodMetricsByMinute[dateId].methods;
+
+  //initialize method
+  if(!methods[method]) {
+    methods[method] = {
+      count: 0,
+      errors: 0,
+      docSize: 0
```
Comment (arunoda) [inline lib/models/methods.js]: Change this into `fetchedDocSize`

**Comment by arunoda** [comment]
@roonyh have you fix the CPU `NaN` issue you mentioned before.
Also try to rebase this one with the master and make this ready to merge.

**Comment by roonyh** [comment]
@arunoda Yeah. that is fixed. Ok. Will rebase.

**Comment by roonyh** [comment]
@arunoda Done.

File: lib/docsize_cache.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  this.values.push(value);
+  this.updated = Date.now();
+
+  if (this.values.length > this.maxValues) {
+    this.values.shift();
+  }
+};
+
+DocSzCacheItem.prototype.getValue = function () {
+  var sorted = this.values.sort();
```
Comment (arunoda) [inline lib/docsize_cache.js]: This sort is wrong. We need to provide a sort function.

File: lib/docsize_cache.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    if(typeof data.get === 'function'){
+      // This is an IdMap
+      data.forEach(function(element){
+        doc = element;
+        return false; // return false to stop loop. We only need one doc.
+      })
+    } else {
+      doc = data[0];
+    }
+    var size = JSON.stringify(doc).length;
```
Comment (arunoda) [inline lib/docsize_cache.js]: Use `Buffer.byteLength()` to get the byte size accurately.

File: lib/hijack/wrap_ddp_stringify.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
+wrapStringifyDDP = function() {
+  var originalStringifyDDP = DDPCommon.stringifyDDP;
+
+  DDPCommon.stringifyDDP = function(msg) {
+    var msgString = originalStringifyDDP(msg);
+    var msgSize = msgString.length;
+
+    var kadiraInfo = Kadira._getInfo(null, true);
```
Comment (arunoda) [inline lib/hijack/wrap_ddp_stringify.js]: Use `Buffer.byteLength`

File: lib/hijack/wrap_ddp_stringify.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      } else if(kadiraInfo.trace.type === 'sub') {
+        Kadira.models.pubsub.trackMsgSize(kadiraInfo.trace.name, "initialSent", msgSize);
+      }
+
+      return msgString;
+    }
+
+    if(Kadira.env.currentSub) {
+      if(Kadira.env.currentSub.__kadiraInfo){
+        Kadira.models.pubsub.trackMsgSize(Kadira.env.currentSub._name, "initialSent", msgSize);
```
Comment (arunoda) [inline lib/hijack/wrap_ddp_stringify.js]: What's the diff between above initialSent and this one?
Are they the same. If so, we could remove the above one.

File: lib/hijack/wrap_observers.js [hunk 15]
```diff
@@ -1,13 +1,22 @@
...
   proto._publishNewResults = function(newResults, newBuffer) {
+    var coll = this._cursorDescription.collectionName;
+    var query = this._cursorDescription.selector;
+    var opts = this._cursorDescription.options;
+    var docSize = Kadira.docSzCache.getSize(coll, query, opts, newResults);
+    var docSize = Kadira.docSzCache.getSize(coll, query, opts, newBuffer);
     var count = newResults.size() + newBuffer.size();
     if(this._ownerInfo) {
       Kadira.models.pubsub.trackPolledDocuments(this._ownerInfo, count);
+      Kadira.models.pubsub.trackDocSize(this._ownerInfo, docSize*data.length);
```
Comment (arunoda) [inline lib/hijack/wrap_observers.js]: This should be `docSize*count` right?

Comment (arunoda) [inline lib/hijack/wrap_observers.js]: Don't we have a title to this docSize?

File: lib/kadira.js [hunk 16]
```diff
@@ -230,3 +231,33 @@ Kadira.trackError = function (type, message, options) {
 Kadira.ignoreErrorTracking = function (err) {
   err._skipKadira = true;
 }
+
+Kadira._postData = function (endpoint, options, callback) {
```
Comment (arunoda) [inline lib/kadira.js]: How do we get this method?

File: lib/models/pubsub.js [hunk 17]
```diff
@@ -327,4 +333,34 @@ PubsubModel.prototype.trackLiveUpdates = function(info, type, count) {
...
+    throw new Error("Kadira: Unknown docs fetched type");
+  }
+}
+
+PubsubModel.prototype.trackMsgSize = function(name, type, size) {
+  var timestamp = Ntp._now();
+  var publication = this._getMetrics(timestamp, name);
+
+  if(type === "liveSent") {
+    publication.fetchedDocSize += size;
```
Comment (arunoda) [inline lib/models/pubsub.js]: this is wrong.

File: lib/models/pubsub.js [hunk 18]
```diff
@@ -327,4 +333,34 @@ PubsubModel.prototype.trackLiveUpdates = function(info, type, count) {
...
+  } else if(type === "initialSent") {
+    publication.initiallyFetchedDocSize += size;
```
Comment (arunoda) [inline lib/models/pubsub.js]: This is wrong.

File: lib/hijack/wrap_ddp_stringify.js [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+      if(kadiraInfo.trace.type === 'method') {
+        Kadira.models.methods.trackMsgSize(kadiraInfo.trace.name, msgSize);
+      } else if(kadiraInfo.trace.type === 'sub') {
+        Kadira.models.pubsub.trackMsgSize(kadiraInfo.trace.name, "initialSent", msgSize);
+      }
+
+      return msgString;
+    }
+
+    if(Kadira.env.currentSub) {
```
Comment (arunoda) [inline lib/hijack/wrap_ddp_stringify.js]: Where do we set `Kadira.env.currentSub` ?

Comment (roonyh) [inline lib/hijack/wrap_ddp_stringify.js]: currentSub is already set here: https://github.com/meteorhacks/kadira/blob/59b1b27035ad4a2ad5a911fad37ec8eb8b7aed40/lib/hijack/wrap_subscription.js#L85

File: lib/docsize_cache.js (hunk 11 above)
Comment (roonyh) [inline lib/docsize_cache.js]: Fixed in 5b14bee

File: lib/docsize_cache.js (hunk 12 above)
Comment (roonyh) [inline lib/docsize_cache.js]: Fixed in 5b14bee

File: lib/kadira.js (hunk 16 above)
Comment (roonyh) [inline lib/kadira.js]: Its probably added mistakenly when rebasing. Removed it.

File: lib/hijack/wrap_observers.js (hunk 15 above)
Comment (roonyh) [inline lib/hijack/wrap_observers.js]: Yes. Fixed both.

File: lib/docsize_cache.js [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+  this.updated = Date.now();
+
+  if (this.values.length > this.maxValues) {
+    this.values.shift();
+  }
+};
+
+DocSzCacheItem.prototype.getValue = function () {
+  function sortNumber(a,b) {
+      return a - b;
```
Comment (arunoda) [inline lib/docsize_cache.js]: 4 spaces?

File: lib/hijack/wrap_ddp_stringify.js (hunk 19 above)
Comment (arunoda) [inline lib/hijack/wrap_ddp_stringify.js]: Okay.

File: package.js [hunk 21]
```diff
@@ -9,7 +9,11 @@ var npmModules = {
   "debug": "0.7.4",
   "kadira-core": "1.3.1",
   "pidusage": "1.0.1",
-  "evloop-monitor": "0.1.0"
+  "evloop-monitor": "0.1.0",
+  "request": "2.51.0",
```
Comment (arunoda) [inline package.js]: No need request.
evloop-monitor is defined twice.

File: lib/hijack/wrap_ddp_stringify.js (hunk 19 above)
Comment (arunoda) [inline lib/hijack/wrap_ddp_stringify.js]: I need  a comment on, how we picked, `Kadira.env.currentSub`.

**Comment by arunoda** [comment]
Once we updated the above one, we are good to go.

**Pull request merged**